File: server/datasets/python_source-code/Multiple-B1/organic_7.py
# Binary Tree Traversal - organic submission 7
# Iterative pre-order, OOP style with __repr__ and a running node count.

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

    def __repr__(self):
        return f"Node({self.val})"


class Tree:
    def __init__(self):
        self.root = None
        self.size = 0

    def insert(self, val):
        self.size += 1
        self.root = self._insert(self.root, val)

    def _insert(self, node, val):
        if node is None:
            return Node(val)
        if val < node.val:
            node.left = self._insert(node.left, val)
        else:
            node.right = self._insert(node.right, val)
        return node

    def preorder_iterative(self):
        if self.root is None:
            return []
        result = []
        stack = [self.root]
        while stack:
            node = stack.pop()
            result.append(node.val)
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        return result


if __name__ == "__main__":
    t = Tree()
    for v in [50, 30, 70, 20, 40, 60, 80]:
        t.insert(v)
    print(f"tree size: {t.size}")
    print("Pre-order (iterative):", t.preorder_iterative())

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_b1/organic_7.java
// Binary Tree Traversal - organic submission 7
// Iterative pre-order, OOP style with running node count.

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class Node {
    int val;
    Node left;
    Node right;

    public Node(int val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

class Tree {
    Node root;
    int size;

    public Tree() {
        this.root = null;
        this.size = 0;
    }

    public void insert(int val) {
        this.size++;
        this.root = insertRec(this.root, val);
    }

    private Node insertRec(Node node, int val) {
        if (node == null) {
            return new Node(val);
        }
        if (val < node.val) {
            node.left = insertRec(node.left, val);
        } else {
            node.right = insertRec(node.right, val);
        }
        return node;
    }

    public List<Integer> preorderIterative() {
        if (this.root == null) {
            return new ArrayList<>();
        }
        List<Integer> result = new ArrayList<>();
        Stack<Node> stack = new Stack<>();
        stack.push(this.root);
        while (!stack.isEmpty()) {
            Node node = stack.pop();
            result.add(node.val);
            if (node.right != null) {
                stack.push(node.right);
            }
            if (node.left != null) {
                stack.push(node.left);
            }
        }
        return result;
    }

    public static void main(String[] args) {
        Tree t = new Tree();
        for (int v : new int[]{50, 30, 70, 20, 40, 60, 80}) {
            t.insert(v);
        }
        System.out.println("tree size: " + t.size);
        System.out.println("Pre-order (iterative): " + t.preorderIterative());
    }
}
